tsk/util/Bitlocker/BitlockerParser.cpp

```diff
@@ -0,0 +1,1565 @@
+* @param keyLen     Will be set to the length of the key
+* 
+* @return SUCCESS on success, GENERAL_ERROR otherwise.
+*/
+BITLOCKER_STATUS BitlockerParser::getKeyData(MetadataEntry* entry, uint8_t** keyDataPtr, size_t& keyLen) {
+
+    // Sanity check
+    if (entry->getValueType() != BITLOCKER_METADATA_VALUE_TYPE::KEY) {
+        writeError("BitlockerParser::getKeyData: Incorrect entry type (" + convertMetadataValueTypeToString(entry->getValueType()) + ")");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    MetadataValueKey* keyValue = dynamic_cast<MetadataValueKey*>(entry->getValue());
+    if (keyValue == nullptr) {
+        writeError("BitlockerParser::getKeyData: Error casting to MetadataValueKey");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Set pointer to the key data
+    *keyDataPtr = keyValue->getKeyBytes();
+    keyLen = keyValue->getKeyLen();
+
+    if (keyDataPtr == nullptr || keyLen == 0) {
+        writeError("BitlockerParser::getKeyData: Key data is invalid");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    return BITLOCKER_STATUS::SUCCESS;
+}
+
+/**
+* Use the decrypted full volume encryption key (FVEK) to initialize the appropriate AES contexts.
+* Also allocate memory for the temp buffers used during decryption.
+* 
+* @param fvekEntry The entry containing the decrypted FVEK
+* 
+* @return SUCCESS on success, GENERAL_ERROR otherwise
+*/
+BITLOCKER_STATUS BitlockerParser::setKeys(MetadataEntry* fvekEntry) {
+
+    if (fvekEntry == nullptr) {
+        writeError("BitlockerParser::setKeys: fvekEntry was null");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    MetadataValueKey* fvek = dynamic_cast<MetadataValueKey*>(fvekEntry->getValue());
+    if (fvek == nullptr) {
+        writeError("BitlockerParser::setKeys: Error casting MetadataValueKey");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Try to initialize the contexts using the encryption type we read from one of the BitLocker headers
+    if (BITLOCKER_STATUS::SUCCESS != setKeys(fvek, m_encryptionType)) {
+
+        // If we failed and the encryption type in the FVEK entry is different than the one we got from the 
+        // header earlier, try again using the encryption type from the FVEK entry
+        if ((fvek->getEncryptionType() != m_encryptionType)
+            && (BITLOCKER_STATUS::SUCCESS == setKeys(fvek, fvek->getEncryptionType()))) {
+
+            // If it worked, change the stored encryption type to this one
+            m_encryptionType = fvek->getEncryptionType();
+            return BITLOCKER_STATUS::SUCCESS;
+        }
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    m_encryptedDataBuffer = (uint8_t*)tsk_malloc(m_sectorSize);
+    m_diffuserTempBuffer = (uint8_t*)tsk_malloc(m_sectorSize);
+    if (m_encryptedDataBuffer == nullptr || m_diffuserTempBuffer == nullptr) {
+        writeError("BitlockerParser::setKeys: Error allocating temp buffers");
```

Comment:
Very unlikely, but the first allocation could succeed and the second could fail. We would then fail to release the memory for the first allocation.

Reply:
I think the destructor would get it later

Reply:
good to resolve.